moveit_ros/trajectory_cache/src/trajectory_cache.cpp

```diff
@@ -0,0 +1,1340 @@
+
+      // Added offset on top of the constraint frame's orientation stated in
+      // the constraint.
+      tf2::Quaternion tf2_quat_goal_offset(constraint.orientation.x, constraint.orientation.y, constraint.orientation.z,
+                                           constraint.orientation.w);
+
+      tf2_quat_frame_offset.normalize();
+      tf2_quat_goal_offset.normalize();
+
+      auto final_quat = tf2_quat_goal_offset * tf2_quat_frame_offset;
+      final_quat.normalize();
+
+      metadata.append(meta_name + ".target_point_offset.x", final_quat.getX());
+      metadata.append(meta_name + ".target_point_offset.y", final_quat.getY());
+      metadata.append(meta_name + ".target_point_offset.z", final_quat.getZ());
+      metadata.append(meta_name + ".target_point_offset.w", final_quat.getW());
+    }
+  }
+
+  return true;
+}
+
+// =============================================================================
+// CARTESIAN TRAJECTORY CACHING
+// =============================================================================
+// CARTESIAN TRAJECTORY CACHING: TOP LEVEL OPS
+moveit_msgs::srv::GetCartesianPath::Request
+TrajectoryCache::construct_get_cartesian_path_request(moveit::planning_interface::MoveGroupInterface& move_group,
+                                                      const std::vector<geometry_msgs::msg::Pose>& waypoints,
+                                                      double step, double jump_threshold, bool avoid_collisions)
+{
+  moveit_msgs::srv::GetCartesianPath::Request out;
+
+  // Some of these parameters need us to pull PRIVATE values out of the
+  // move_group elsewhere... Yes, it is very cursed and I hate it.
+  // Fixing it requires fixing it in MoveIt.
+  moveit_msgs::msg::MotionPlanRequest tmp;
+  move_group.constructMotionPlanRequest(tmp);
+
+  out.start_state = std::move(tmp.start_state);
+  out.group_name = std::move(tmp.group_name);
+  out.max_velocity_scaling_factor = tmp.max_velocity_scaling_factor;
+  out.max_acceleration_scaling_factor = tmp.max_acceleration_scaling_factor;
+
+  out.header.frame_id = move_group.getPoseReferenceFrame();
+  out.waypoints = waypoints;
+  out.max_step = step;
+  out.jump_threshold = jump_threshold;
+  out.path_constraints = moveit_msgs::msg::Constraints();
+  out.avoid_collisions = avoid_collisions;
+  out.link_name = move_group.getEndEffectorLink();
+
+  // We were already cursed before, now we are double cursed...
+  // move_group.getNodeHandle() is UNIMPLEMENTED upstream.
+  out.header.stamp = node_->now();
+
+  return out;
+}
+
+std::vector<MessageWithMetadata<moveit_msgs::msg::RobotTrajectory>::ConstPtr>
+TrajectoryCache::fetch_all_matching_cartesian_trajectories(
+    const moveit::planning_interface::MoveGroupInterface& move_group, const std::string& move_group_namespace,
+    const moveit_msgs::srv::GetCartesianPath::Request& plan_request, double min_fraction, double start_tolerance,
```

Comment:
Can you make even more arguments const here?

Reply:
The remaining arguments are values.

Clang suggests to not do so.
https://clang.llvm.org/extra/clang-tidy/checks/readability/avoid-const-params-in-decls.html